src/common/client_test.go

```diff
@@ -0,0 +1,153 @@
+				return nil, fmt.Errorf("should not be called")
+			},
+			CreateAliasHandler: func(input opslevel.AliasCreateInput) error {
+				return fmt.Errorf("should not be called")
+			},
+			AssignTagsHandler: func(service *opslevel.Service, tags map[string]string) error {
+				return fmt.Errorf("should not be called")
+			},
+			AssignPropertyHandler: func(input opslevel.PropertyInput) error {
+				return fmt.Errorf("should not be called")
+			},
+			CreateTagHandler: func(input opslevel.TagCreateInput) error {
+				return fmt.Errorf("should not be called")
+			},
+			CreateToolHandler: func(tool opslevel.ToolCreateInput) error {
+				return fmt.Errorf("should not be called")
+			},
+			GetRepositoryWithAliasHandler: func(alias string) (*opslevel.Repository, error) {
+				return nil, fmt.Errorf("should not be called")
+			},
+			CreateServiceRepositoryHandler: func(input opslevel.ServiceRepositoryCreateInput) error {
+				return fmt.Errorf("should not be called")
+			},
+			UpdateServiceRepositoryHandler: func(input opslevel.ServiceRepositoryUpdateInput) error {
+				return fmt.Errorf("should not be called")
+			},
+		},
+	}
+}
+
+// NewTestClientBuilderWithPanic creates a client where every function that has not been explicitly set will panic.
+func NewTestClientBuilderWithPanic() *TestClientBuilder {
+	return &TestClientBuilder{
+		client: common.OpslevelClient{
+			GetServiceHandler: func(alias string) (*opslevel.Service, error) {
+				panic("should not be called")
+			},
+			CreateServiceHandler: func(input opslevel.ServiceCreateInput) (*opslevel.Service, error) {
+				panic("should not be called")
+			},
+			UpdateServiceHandler: func(input opslevel.ServiceUpdateInput) (*opslevel.Service, error) {
+				panic("should not be called")
+			},
+			CreateAliasHandler: func(input opslevel.AliasCreateInput) error {
+				panic("should not be called")
+			},
+			AssignTagsHandler: func(service *opslevel.Service, tags map[string]string) error {
+				panic("should not be called")
+			},
+			AssignPropertyHandler: func(input opslevel.PropertyInput) error {
+				panic("should not be called")
+			},
+			CreateTagHandler: func(input opslevel.TagCreateInput) error {
+				panic("should not be called")
+			},
+			CreateToolHandler: func(tool opslevel.ToolCreateInput) error {
+				panic("should not be called")
+			},
+			GetRepositoryWithAliasHandler: func(alias string) (*opslevel.Repository, error) {
+				panic("should not be called")
+			},
+			CreateServiceRepositoryHandler: func(input opslevel.ServiceRepositoryCreateInput) error {
+				panic("should not be called")
+			},
+			UpdateServiceRepositoryHandler: func(input opslevel.ServiceRepositoryUpdateInput) error {
+				panic("should not be called")
+			},
+		},
+	}
+}
+
+func (b *TestClientBuilder) SetAssignPropertyHandler(fn func(input opslevel.PropertyInput) error) *TestClientBuilder {
+	b.client.AssignPropertyHandler = fn
+	return b
```

Comment:
Rather than returning the `TestClientBuilder` with each member function, why not just call all the `SetThingHandler()` functions first then pass in the configured `TestClientBuilder` after?

Reply:
@davidbloss If you mean this,

```
b := Builder()
b.SetABC()
b.SetXYZ()
b.Build() // returns wanted object
```

The disadvantage of this is that it requires multiple statements whereas when you return the builder in each set call you can do everything on one line.

This is necessary for instantiating clients in places like test case definitions, for example:

```
		"Happy Path": {
			registration: testRegistration,
			reconciler: common.NewServiceReconciler(NewTestClientBuilderWithError().SetGetServiceHandler(func(alias string) (*opslevel.Service, error) {
				return &opslevel.Service{}, nil			}).SetCreateServiceHandler(func(input opslevel.ServiceCreateInput) (*opslevel.Service, error) {
				return &testService, nil
			}).GetClient(), false),
			assert: func(t *testing.T, err error) {
				autopilot.Ok(t, err)
			},
		},
```

Reply:
In your example you wouldn't need to call `b.Build()` since `b` would already have everything set.

`b.client` can only be set after calling `NewTestClientBuilderWith...()`:
- here we have `b := Builder()`
- all the following `b.SetABC()`,  `b.SetXYZ()` calls would set the client's handler fields
- then you can pass in `b` with any final `Build()` 

My point is we don't need to return `*TestClientBuilder` in each member function.